File: agenttel-node/src/agentic/tracer.ts
import { trace, type Span, type Tracer, SpanKind } from '@opentelemetry/api';
import * as ag from '../agentic-attributes.js';
import { AgentType, InvocationStatus, type StepType } from '../enums-impl.js';

export interface AgentTracerOptions {
  agentName: string;
  agentType?: AgentType;
  framework?: string;
  agentVersion?: string;
}

/** Creates agent invocation traces with step/task scopes. */
export class AgentTracer {
  private readonly tracer: Tracer;
  private readonly agentName: string;
  private readonly agentType: AgentType;
  private readonly framework?: string;
  private readonly agentVersion?: string;

  constructor(options: AgentTracerOptions, tracerOrName?: Tracer | string) {
    this.tracer = typeof tracerOrName === 'string'
      ? trace.getTracer(tracerOrName)
      : tracerOrName ?? trace.getTracer('agenttel.agentic');
    this.agentName = options.agentName;
    this.agentType = options.agentType ?? AgentType.SINGLE;
    this.framework = options.framework;
    this.agentVersion = options.agentVersion;
  }

  /** Start a new agent invocation. */
  invoke(goal: string): AgentInvocation {
    const span = this.tracer.startSpan('invoke_agent', { kind: SpanKind.INTERNAL });

    span.setAttribute(ag.AGENTIC_AGENT_NAME, this.agentName);
    span.setAttribute(ag.AGENTIC_AGENT_TYPE, this.agentType);
    span.setAttribute(ag.AGENTIC_INVOCATION_GOAL, goal);
    span.setAttribute(ag.AGENTIC_INVOCATION_STATUS, InvocationStatus.RUNNING);
    if (this.framework) span.setAttribute(ag.AGENTIC_AGENT_FRAMEWORK, this.framework);
    if (this.agentVersion) span.setAttribute(ag.AGENTIC_AGENT_VERSION, this.agentVersion);

    return new AgentInvocation(span, this.tracer);
  }
}

/** Represents an agent invocation scope. */
export class AgentInvocation {
  private readonly span: Span;
  private readonly tracer: Tracer;
  private stepCount = 0;

  constructor(span: Span, tracer: Tracer) {
    this.span = span;
    this.tracer = tracer;
  }

  /** Create a new reasoning step. */
  step(stepType: StepType, description?: string): StepScope {
    this.stepCount++;
    const span = this.tracer.startSpan('agenttel.agentic.step', { kind: SpanKind.INTERNAL });
    span.setAttribute(ag.AGENTIC_STEP_NUMBER, this.stepCount);
    span.setAttribute(ag.AGENTIC_STEP_TYPE, stepType);
    return new StepScope(span);
  }

  /** Create a new task scope. */
  task(taskName: string): TaskScope {
    const span = this.tracer.startSpan('agenttel.agentic.task', { kind: SpanKind.INTERNAL });
    span.setAttribute(ag.AGENTIC_TASK_NAME, taskName);
    span.setAttribute(ag.AGENTIC_TASK_STATUS, 'in_progress');
    span.setAttribute(ag.AGENTIC_TASK_DEPTH, 0);
    return new TaskScope(span, this.tracer, 0);
  }

  /** Complete with success. */
  end(): void {
    this.span.setAttribute(ag.AGENTIC_INVOCATION_STATUS, InvocationStatus.COMPLETED);
    this.span.setAttribute(ag.AGENTIC_INVOCATION_STEPS, this.stepCount);
    this.span.end();
  }

  /** Complete with error. */
  endError(error: Error): void {
    this.span.setAttribute(ag.AGENTIC_INVOCATION_STATUS, InvocationStatus.FAILED);
    this.span.setAttribute(ag.AGENTIC_INVOCATION_STEPS, this.stepCount);
    this.span.recordException(error);
    this.span.end();
  }

  getSpan(): Span { return this.span; }
}

/** A single reasoning step scope. */
export class StepScope {
  private readonly span: Span;

  constructor(span: Span) { this.span = span; }

  toolName(name: string): this { this.span.setAttribute(ag.AGENTIC_STEP_TOOL_NAME, name); return this; }
  toolStatus(status: string): this { this.span.setAttribute(ag.AGENTIC_STEP_TOOL_STATUS, status); return this; }
  error(err: Error): void { this.span.recordException(err); }
  end(): void { this.span.end(); }
  getSpan(): Span { return this.span; }
}

/** A task or subtask scope. */
export class TaskScope {
  private readonly span: Span;
  private readonly tracer: Tracer;
  private readonly depth: number;

  constructor(span: Span, tracer: Tracer, depth: number) {
    this.span = span;
    this.tracer = tracer;
    this.depth = depth;
  }

  subtask(name: string): TaskScope {
    const span = this.tracer.startSpan('agenttel.agentic.task', { kind: SpanKind.INTERNAL });
    span.setAttribute(ag.AGENTIC_TASK_NAME, name);
    span.setAttribute(ag.AGENTIC_TASK_STATUS, 'in_progress');
    span.setAttribute(ag.AGENTIC_TASK_DEPTH, this.depth + 1);
    return new TaskScope(span, this.tracer, this.depth + 1);
  }

  complete(): void { this.span.setAttribute(ag.AGENTIC_TASK_STATUS, 'completed'); }
  fail(reason: string): void {
    this.span.setAttribute(ag.AGENTIC_TASK_STATUS, 'failed');
    this.span.setAttribute('agenttel.agentic.task.failure_reason', reason);
  }
  end(): void { this.span.end(); }
  getSpan(): Span { return this.span; }
}

File: agenttel-go/agentic/tracer.go
// Package agentic provides agentic observability: agent tracing, scopes, and orchestration patterns.
package agentic

import (
	"context"
	"sync/atomic"

	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/trace"

	agattr "go.agenttel.dev/agenttel/attributes"
	"go.agenttel.dev/agenttel/enums"
)

// AgentTracer creates agent invocation traces.
type AgentTracer struct {
	tracer       trace.Tracer
	agentName    string
	agentType    enums.AgentType
	framework    string
	agentVersion string
}

// TracerBuilder builds an AgentTracer.
type TracerBuilder struct {
	tracer       trace.Tracer
	agentName    string
	agentType    enums.AgentType
	framework    string
	agentVersion string
}

// NewTracer creates a TracerBuilder.
func NewTracer(tracer trace.Tracer) *TracerBuilder {
	return &TracerBuilder{tracer: tracer, agentType: enums.AgentTypeSingle}
}

// AgentName sets the agent name.
func (b *TracerBuilder) AgentName(name string) *TracerBuilder { b.agentName = name; return b }

// AgentType sets the agent type.
func (b *TracerBuilder) AgentType(t enums.AgentType) *TracerBuilder { b.agentType = t; return b }

// Framework sets the agent framework.
func (b *TracerBuilder) Framework(f string) *TracerBuilder { b.framework = f; return b }

// Version sets the agent version.
func (b *TracerBuilder) Version(v string) *TracerBuilder { b.agentVersion = v; return b }

// Build creates the AgentTracer.
func (b *TracerBuilder) Build() *AgentTracer {
	return &AgentTracer{
		tracer:       b.tracer,
		agentName:    b.agentName,
		agentType:    b.agentType,
		framework:    b.framework,
		agentVersion: b.agentVersion,
	}
}

// Invoke starts a new agent invocation span.
func (t *AgentTracer) Invoke(ctx context.Context, goal string) (context.Context, *Invocation) {
	ctx, span := t.tracer.Start(ctx, "invoke_agent", trace.WithSpanKind(trace.SpanKindInternal))

	span.SetAttributes(
		attribute.String(agattr.AgenticAgentName, t.agentName),
		attribute.String(agattr.AgenticAgentType, string(t.agentType)),
		attribute.String(agattr.AgenticInvocationGoal, goal),
		attribute.String(agattr.AgenticInvocationStatus, string(enums.InvocationStatusRunning)),
	)
	if t.framework != "" {
		span.SetAttributes(attribute.String(agattr.AgenticAgentFramework, t.framework))
	}
	if t.agentVersion != "" {
		span.SetAttributes(attribute.String(agattr.AgenticAgentVersion, t.agentVersion))
	}

	return ctx, &Invocation{
		ctx:    ctx,
		span:   span,
		tracer: t.tracer,
	}
}

// Invocation represents an agent invocation scope.
type Invocation struct {
	ctx       context.Context
	span      trace.Span
	tracer    trace.Tracer
	stepCount atomic.Int64
}

// Step creates a new reasoning step within the invocation.
func (inv *Invocation) Step(ctx context.Context, stepType enums.StepType, description string) (context.Context, *StepScope) {
	num := inv.stepCount.Add(1)
	ctx, span := inv.tracer.Start(ctx, "agenttel.agentic.step", trace.WithSpanKind(trace.SpanKindInternal))

	span.SetAttributes(
		attribute.Int64(agattr.AgenticStepNumber, num),
		attribute.String(agattr.AgenticStepType, string(stepType)),
	)

	return ctx, &StepScope{span: span, description: description}
}

// Task creates a new task scope within the invocation.
func (inv *Invocation) Task(ctx context.Context, taskName string) (context.Context, *TaskScope) {
	ctx, span := inv.tracer.Start(ctx, "agenttel.agentic.task", trace.WithSpanKind(trace.SpanKindInternal))

	span.SetAttributes(
		attribute.String(agattr.AgenticTaskName, taskName),
		attribute.String(agattr.AgenticTaskStatus, "in_progress"),
		attribute.Int64(agattr.AgenticTaskDepth, 0),
	)

	return ctx, &TaskScope{ctx: ctx, span: span, tracer: inv.tracer, depth: 0}
}

// End completes the invocation with success status.
func (inv *Invocation) End() {
	inv.span.SetAttributes(
		attribute.String(agattr.AgenticInvocationStatus, string(enums.InvocationStatusCompleted)),
		attribute.Int64(agattr.AgenticInvocationSteps, inv.stepCount.Load()),
	)
	inv.span.End()
}

// EndError completes the invocation with error status.
func (inv *Invocation) EndError(err error) {
	inv.span.SetAttributes(
		attribute.String(agattr.AgenticInvocationStatus, string(enums.InvocationStatusFailed)),
		attribute.Int64(agattr.AgenticInvocationSteps, inv.stepCount.Load()),
	)
	inv.span.RecordError(err)
	inv.span.End()
}

// Span returns the underlying OTel span.
func (inv *Invocation) Span() trace.Span { return inv.span }
